kratos/solving_strategies/response_functions/adjoint_response_function.h

```diff
@@ -0,0 +1,286 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:
+//
+
+#ifndef ADJOINT_RESPONSE_FUNCTION_H
+#define ADJOINT_RESPONSE_FUNCTION_H
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "includes/define.h"
+#include "includes/element.h"
+#include "includes/condition.h"
+#include "includes/process_info.h"
+#include "includes/ublas_interface.h"
+#include "containers/array_1d.h"
+#include "containers/variable.h"
+
+
+// Application includes
+
+namespace Kratos
+{
+///@name Kratos Classes
+///@{
+
+/// A base class for adjoint response functions.
+class KRATOS_API(KRATOS_CORE) AdjointResponseFunction
+{
+public:
+    ///@name Type Definitions
+    ///@{
+
+    KRATOS_CLASS_POINTER_DEFINITION(AdjointResponseFunction);
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /// Constructor.
+    AdjointResponseFunction()
+    {
+    }
+
+    /// Destructor.
+    virtual ~AdjointResponseFunction()
+    {
+    }
+
+    ///@}
+    ///@name Operators
+    ///@{
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    virtual void Initialize()
+    {
+    }
+
+    virtual void InitializeSolutionStep()
+    {
+    }
+
+    virtual void FinalizeSolutionStep()
+    {
+    }
+
+    /// Calculate the local gradient w.r.t. primal solution.
+    /**
+     * @param[in]     rAdjointElement    the adjoint element.
+     * @param[in]     rResidualGradient  the transposed gradient of the
+     *                                   element's residual w.r.t. primal.
+     * @param[out]    rResponseGradient  the gradient of the response function.
+     * @param[in]     rProcessInfo       the current process info.
+     */
+    virtual void CalculateGradient(const Element& rAdjointElement,
+                                   const Matrix& rResidualGradient,
+                                   Vector& rResponseGradient,
+                                   ProcessInfo& rProcessInfo)
+    {
+        KRATOS_ERROR << "Calling base class response function method.\n";
```

Comment:
Why do you distinguish between these functions with this error message and the pure virtual function ```CalculateValue```?
From a developer point of you you are forced to implement both, but for one case you notice at run time, for the other case at compile time.

Reply:
Yes, that is the only distinction. With pure virtual we are making `CalculateValue` mandatory. The other methods are optional, provided you don't try to use one of them. For example, you might use `CalculateGradient` for a static response, but not `CalculateSecondDerivativesGradient`. As long as you don't try passing your static response to a client which expects second derivatives gradient you will be fine.